NEW Show region in company info & Global option to show state code MAIN_SHOW_STATE_CODE

ruistrecht-alt opened the pull request:
# NEW Show region in company info & Global option to show state code MAIN_SHOW_STATE_CODE
Added new option MAIN_SHOW_STATE_CODE (on/off; 1/0) to show state code in select fields options and values.
Changed option MAIN_SHOW_REGION_IN_STATE to on/off - 1/0 and to mix with the new MAIN_SHOW_STATE_CODE (can show region name and state code)

ruistrecht-alt committed:
Added new option MAIN_SHOW_STATE_CODE (on/off; 1/0) to show state code in select fields options and values.

Changed option MAIN_SHOW_REGION_IN_STATE to on/off - 1/0 and to mix with the new MAIN_SHOW_STATE_CODE (can show region name and state code)

ruistrecht-alt committed:
Added self to contributors list

eldy commented on htdocs/core/lib/company.lib.php:
```diff
@@ -390,14 +391,17 @@ function getCountry($searchkey,$withcode='',$dbtouse=0,$outputlangs='',$entconv=
  *    @param	DoliDB		$dbtouse		Database handler (using in global way may fail because of conflicts with some autoload features)
  *    @return   string      				String with state code or state name (Return value is always utf8 encoded and without entities)
  */
-function getState($id,$withcode='',$dbtouse=0)
```
To avoid breaking compatiblity, we must not remove a parameter on a function.
Also, getState is a lib function that must be stateless, so behaviour, showing state or not and region or not, is better if it depends on an entry parameter, not on an external setup.
Can you change code so function getState is

function getState($id,$withcode='',$dbtouse=0, $withregion='')

ruistrecht-alt replied:
Ok, I'm working on it. 
Just a question. What is the purpose of the second parameter on the function call: 
$mysoc->state=getState($mysoc->state_code,0);
Located in commondocgenerator.class.php, Line 108?
$withcode is a string that has no logic built for the integer value 0.

ruistrecht-alt commented on htdocs/core/lib/company.lib.php:
```diff
@@ -406,11 +410,22 @@ function getState($id,$withcode='',$dbtouse=0)
-            $label=$obj->label;
-            if ($withcode == '1') return $label=$obj->code?"$obj->code":"$obj->code - $label";
-            else if ($withcode == '2') return $label=$obj->code;
-            else if ($withcode == 'all') return array('id'=>$obj->rowid,'code'=>$obj->code,'label'=>$label);
```
Also, $withcode == 'all' is not used anywhere on the rest of the code. Should I leave it alone? 
Showing rowid in a select field seems kind of pointless to me

ruistrecht-alt replied:
I found it. 'all' as not been implemented yet in getState(), the behavior should be like getCountry():
$tmparray=getCountry(GETPOST('country_id','int'),'all',$db,$langs,0);
	if (! empty($tmparray['id']))
	{
		$mysoc->country_id   =$tmparray['id'];
		$mysoc->country_code =$tmparray['code'];
		$mysoc->country_label=$tmparray['label'];
		$s=$mysoc->country_id.':'.$mysoc->country_code.':'.$mysoc->country_label;
		dolibarr_set_const($db, "MAIN_INFO_SOCIETE_COUNTRY", $s,'chaine',0,'',$conf->entity);
	}

Right?

ruistrecht-alt committed:
Updated method to fetch state name, code and id to match method used with country

Reverted some previous changes to allow backwards compatibility
Extended MAIN_SHOW_REGION_IN_STATE to include showing region in thirdparty card

ruistrecht-alt committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt committed:
Added self to contributors list

ruistrecht-alt committed:
Added missing comments for new function parameters, removed unused parameter $searchlabel

ruistrecht-alt committed:
Tiny fix to php doc variables

ruistrecht-alt committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt commented:
PR as been updated

ruistrecht-alt committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

# Conflicts:
#	htdocs/admin/company.php
#	htdocs/core/class/commonobject.class.php

ruistrecht-alt committed:
Fixed issue with merge error

ruistrecht-alt committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

# Conflicts:
#	htdocs/societe/class/societe.class.php

ruistrecht-alt commented:
@eldy PR conflicts have been fixed and compatibility issues have been worked around

ruistrecht-alt committed:
Merge branch 'develop' of https://github.com/Dolibarr/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt commented:
@eldy Is this PR dead? Does it have any pending issue?

ruistrecht-alt committed:
Merge branch 'develop' into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt committed:
Merge branch 'NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE' of https://github.com/ruistrecht-alt/dolibarr into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt commented:
PR has been updated to reflect changes made in PR #8100

ruistrecht-alt committed:
Forgot to update other files to same PR changes

eldy commented:
For perfomance reason, we must no have any code that include sql request into method  setMysoc
It seems goal of this request is to be able to have the region stored inside the setup ?
We can just add it into MAIN_INFO_SOCIETE_REGION like we do for the state into 
MAIN_INFO_SOCIETE_STATE, when we save the setup of company ?

ruistrecht-alt commented:
@eldy you misunderstood the purpose of this PR. 
The goal of this PR is to add a way to show the Region, alongside the state, in the company info.
It seems I made a mistake in an earlier commit and I'll update soon

ruistrecht-alt committed:
Merge branch 'develop' into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

# Conflicts:
#	htdocs/core/class/commonobject.class.php

ruistrecht-alt committed:
Trying to not include sql request into method setMysoc

ruistrecht-alt committed:
Forgot withregion condition if we want no state code

ruistrecht-alt committed:
Forgot that MAIN_SHOW_STATE_CODE is not just 0 or 1

ruistrecht-alt committed:
Added translation to company.php if MAIN_SHOW_REGION_IN_STATE_SELECT is activated

ruistrecht-alt committed:
Fixed wrong global

ruistrecht-alt committed:
Merge branch 'develop' into NEW_Show_Region_in_company_info_&_Global_option_to_show_state_code_MAIN_SHOW_STATE_CODE

ruistrecht-alt commented:
@eldy updated commit to exclude sql request in method setMysoc. 
Need a reviewer to confirm if my update is consistent or if I missed something

eldy committed:
Update company.lib.php
